kythe/cxx/extractor/cxx_extractor.cc

```diff
@@ -909,62 +910,27 @@ class ExtractorAction : public clang::PreprocessorFrontendAction {
 
 }  // anonymous namespace
 
-void KindexWriterSink::OpenIndex(const std::string& hash) {
-  using namespace google::protobuf::io;
-  CHECK(open_path_.empty() && fd_ < 0)
-      << "Reopening a KindexWriterSink (old fd:" << fd_
-      << " old path: " << open_path_ << ")";
-  std::string file_path = single_file_ ? path_ : path_ + "/" + hash + ".kindex";
-  // Open with read/write for all. The exact permissions required may depend on
-  // the backend implementation running the extractor.
-  fd_ = ::open(file_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0666);
-  CHECK_GE(fd_, 0) << "Couldn't open output file " << file_path;
-  open_path_ = file_path;
-  file_stream_ = absl::make_unique<FileOutputStream>(fd_);
-  GzipOutputStream::Options options;
-  // Accept the default compression level and compression strategy.
-  options.format = GzipOutputStream::GZIP;
-  gzip_stream_ =
-      absl::make_unique<GzipOutputStream>(file_stream_.get(), options);
-  coded_stream_ = absl::make_unique<CodedOutputStream>(gzip_stream_.get());
-}
-
-KindexWriterSink::~KindexWriterSink() {
-  CHECK(!(coded_stream_ && coded_stream_->HadError()))
-      << "Errors encountered writing to " << open_path_;
-  coded_stream_.reset();
-  gzip_stream_.reset();
-  file_stream_.reset();
-  close(fd_);
-}
+KzipWriterSink::KzipWriterSink(const std::string& path, bool single_file)
+    : path_(path), single_file_(single_file) {}
 
-void KindexWriterSink::WriteHeader(
-    const kythe::proto::CompilationUnit& header) {
-  coded_stream_->WriteVarint32(header.ByteSizeLong());
-  CHECK(header.SerializeToCodedStream(coded_stream_.get()))
-      << "Couldn't write header to " << open_path_;
+void KzipWriterSink::OpenIndex(const std::string& unit_hash) {
+  CHECK(writer_ == nullptr) << "OpenIndex() called twice";
+  std::string path =
+      single_file_ ? path_ : JoinPath(path_, unit_hash + ".kzip");
+  writer_ = absl::make_unique<IndexWriter>(OpenKzipWriterOrDie(path));
```

Comment:
nit: absl::optional<IndexWriter> rather than unique_ptr? (IndexWriter is itself a pimpl class).